recommendation-app/README.md

```diff
@@ -1,41 +1,338 @@
+```
+
+Now we have a view which, for each clickstream event, emits a list of in-stock products in the same category and also marks those purchased by the user.  
+However, these are in no particular order.
+To indicate how good of a recommendation each entry might be, we can use the `ROW NUMBER` [aggregate function](https://nightlies.apache.org/flink/flink-docs-stable/docs/dev/table/functions/systemfunctions/#aggregate-functions). 
+This will label each item sequentially based on the ordering within the window of items currently being processed.
+We can use the [`OVER`](https://nightlies.apache.org/flink/flink-docs-stable/docs/dev/table/sql/queries/over-agg/#over-aggregation) clause combined with a `PARTITION BY` to group results first into partitions by user ID and then `ORDER BY` to order those partitioned products first by whether they have been purchased previously and then by their average rating.
 
-Output: A stream of recommendations (user id, 6 product ids)
+The statement below illustrates the operations described above:
+
+```sql
+CREATE TEMPORARY VIEW top_products AS 
+  SELECT cp.user_id, 
+    cp.event_time, 
+    cp.product_id, 
+    cp.category, 
+    cp.stock, 
+    cp.rating, 
+    cp.purchased, 
+    ROW_NUMBER() OVER (PARTITION BY cp.user_id ORDER BY cp.purchased DESC, cp.rating DESC) AS rn 
+  FROM category_products cp
+; 
+```
+
+Now we have a ranked list of products per user clickstream event.
+To recommend the top 6, we simply need to `SELECT` from the view above using a `WHERE rn <= 6` clause and `GROUP BY` the user ID.
+To package the results up more cleanly we can use the `LISTAGG` [aggregation function](https://nightlies.apache.org/flink/flink-docs-stable/docs/dev/table/functions/systemfunctions/#aggregate-functions) to combine the product IDs into a single list.
+We can also define over what time window we want to source data for the recommendation by defining a [window aggregation](https://nightlies.apache.org/flink/flink-docs-stable/docs/dev/table/sql/queries/window-agg/#group-window-functions) clause below the `GROUP BY` clause. 
+The simplest time window is a tumbling window, which collects data over a defined period, performs the aggregation and then moves forward by the same period and so on.
+The longer the window, the more clickstream events will be included but the slower the results will be to update and more data you need to aggregate.
+
+A statement to perform the final recommendation query is shown below:
+
+```sql
+SELECT 
+  user_id, 
+  LISTAGG(product_id, ',') AS top_product_ids, 
+  TUMBLE_END(event_time, INTERVAL '5' SECOND) as window_end_timestamp
+FROM top_products 
+WHERE rn <= 6 
+GROUP BY user_id, 
+TUMBLE(event_time, INTERVAL '5' SECOND)
+;
+```
+
+### Saving the results
+
+Now that we can generate the product recommendations we need to save that data somewhere. 
+The easiest option is to output the result back to Kafka so that some other service can pick them up and present them to the user.
```

Comment:
```suggestion
The easiest option is to output the results back to Kafka so that some other service can pick them up and present them to the user.
```